### ISSUE TITLE
No Hook available for prompts on opening a file

### ISSUE THREAD

**Issue by purxiz** [mawww/kakoune#4102]
I'd like to have a script be run to check lock files whenever I try to edit (i.e. with `:e` or `:edit`) a file, and then prompt me to overwrite the lock if it exists. I've gotten this to work without issue if I run a command like `:checklock`, but if I try to add a hook on BufCreate, the prompt from my command never appears. I can verify that the script is running by making it output to a log file, but there doesn't seem to a hook that executes on opening a file (like BufCreate), but late enough in the file opening process that `prompt,` `info,` etc. can display information as a result of the hook.

In short, I'd like to be able to do 
```
hook global BufCreate .+ %{
    prompt do_something? %{do something else}
}
```
but it doesn't seem to be currently possible.

**Comment by Screwtapello** [comment]
`Buf*` hooks are always executed in buffer context. You can have more buffers open than there are windows to display them in, so commands that try to interact with the user (like `:prompt`, `:info`, etc.) do nothing or outright fail in buffer context.

In order to interact with the user, you need a window, so you need a hook that executes in window context. Of the various `Win*` hooks, most of them say "This hook is executed in a draft context", rather than a window context, so they can't interact with the user either. The only remaining hook is `WinDisplay`, which runs the first time a given buffer is displayed in a given window, so that's probably the one you want to use.

**Comment by purxiz** [comment]
Ah, thank you! The documentation for that was very unclear to me. I'll try it out tomorrow and close the issue if it works.

**Comment by caksoylar** [comment]
For your second edit, you can use this trick (I got the idea from @Screwtapello originally too, IIRC):
```kak
hook global WinCreate .* %{
    hook -once window WinDisplay .* %{
        # here
    }
}
```

**Comment by Screwtapello** [comment]
I think that inner hook needs to be `hook -once buffer WinDisplay ...` since (as mentioned) `BufCreate` hooks execute in buffer context, so they don't have a window, so you can't install hooks (or mappings, or highlighters, or whatever) at window context because there's no window yet.

If you install `WinDisplay` in buffer scope, it will automatically trigger for every window that ever displays the buffer. If you install the hook with `-once`, it will automatically trigger for the *first* window that displays the buffer, and never again, which I think is what you want.
